### ISSUE TITLE
navigation.goBack() to previous page is impossible if that page was on another stackNavigator

### ISSUE THREAD

**Issue by Bonobomagno** [react-navigation/react-navigation#4403]
Update: added snack and example

### Your Environment

| software         | version
| ---------------- | -------
| react-navigation | 2.0.x
| react-native     | expo 27
| node             | 9.11.1
| npm or yarn      | yarn 1.7

```
tab:
   stack1:
      screen1,
      detailScreen1
      (default route screen1)
   stack2:
      screen2,
      detailScreen2
    (default route screen2)

```
so, for eg, from Screen1 i call 
`navigation.navigate('detailScreen2')`

using on detailScreen2
` headerLeft: <GoBack onPress={() => navigation.goBack(null)} />`
move me to Screen2 instead of screen1 .
In fact, goBack call stack2 (who call screen2 as default screen for the stack).

How to fix this? Because if i call `navigation.navigate('screen1')`, the screen stack will be still full, so when i will press on stack2 tab icon, the new screen will be detailScreen2 instead of screen2 (because of the old opened screen on stack2)

Update:
[snack](https://snack.expo.io/ryuU43bg7)

Easy example made from navigation playground.
On home Tab go to Profile page, after go to notification, then go back.
The back will move you to the true parent (settings screen) instead to the previous page (profile screen).

**Comment by matthamil** [comment]
Hi! Can you create a reproducible example with [Snack](https://snack.expo.io)? Can you also follow the [issue template](https://github.com/react-navigation/react-navigation/blob/master/.github/ISSUE_TEMPLATE.md)? Thanks!

**Comment by Bonobomagno** [comment]
updated

**Comment by Bonobomagno** [comment]
Possible Solution:
``    const navi = (routeBack) => {
      navigation.popToTop()
      navigation.navigate(routeBack)
    }``

Calling popToTop() before navigate, will clear the actual tab stack, and move the the old (routBack, passed as param) without drawback.

**Comment by Bonobomagno** [comment]
Also, there is a graphical glitch when navigating from detailscreen1 to detailscreen2.
When doing that, the transaction show scree2n for a small timeframe and detailscreen2 after.
Not so cool to have...

**Comment by brentvatne** [comment]
hey @Bonobomagno - I don't quite understand what you're trying to explain here. can you describe it in the following format please?

```
- In the provided Snack, press X
- Next, press Y
- ... etc
- Notice that Z happens. I would expect A to happen.
```

**Comment by brentvatne** [comment]
```
Easy example made from navigation playground.
On home Tab go to Profile page, after go to notification, then go back.
The back will move you to the true parent (settings screen) instead to the previous page (profile screen).
```

I believe this is expected - `goBack` will act on the deepest active router, which in this case is the stack. if you'd like to go back to the other tab you are right that you would need to use `navigate`. cc @ericvicenti to confirm as I was never totally clear on the purpose of `goBack(null)`

**Comment by brentvatne** [comment]
<img width="1111" alt="screen shot 2018-06-11 at 3 07 22 pm" src="https://user-images.githubusercontent.com/90494/41259655-37a25d2e-6d89-11e8-9686-8ad5721e6144.png">

@ericvicenti went on to add:

```
but as far as chronological back.. its not really baked into the current understanding of navigation state
we could potentially add a `state.lastActiveKey` or something, or even an array of previously active keys.. but. the chronological thing doesn't make perfect sense... I mean, do you ever want to go chronologically back from a 'pop'? It would effectively be a push
(and an array of previously active route keys wouldn't even help there...)
```

tl;dr: this is working as expected. the docs need some better wording around `goBack(null)`. if you would like to propose introducing another way of handling back actions, please submit a rfc with some motivating use cases: https://github.com/react-navigation/rfcs

another way you could handle this in your app is to pass the route info for the route you want to return to in the params when navigating to another route. then check if that param exists and go back there if so.

**Comment by Bonobomagno** [comment]
Thanks for the reply! 
In this case, a better wording would be cool. At least for understand what goBack() does.

If you use navigate(), the stack will not pop che old tab stack.
So my actual working solution is
`navigation.popToTop();  navigation.navigate(routeBack) `

Sometimes it glitch, showing both popToTop and navigate animation, but this is the only working one.
Maybe a popToRoute(routeBack) function, with only 1 animation, would be nicer to have.

**Comment by uladkasach** [comment]
@brentvatne is there a recommended solution for going "chronologically back" (i.e., "take me back to the last screen i was on") since `goBack()` currently only goes "hierarchically back" (i.e., "take me back back through the navigational hierarchy" - per your screenshotted conversation with `evv`)?

**Comment by matthamil** [comment]
@uladkasach Sounds like [`navigation.pop()`](https://reactnavigation.org/docs/en/stack-actions.html#pop) might be what you are looking for?

**Comment by toancaro** [comment]
Change to `navigation.pop()` did not work. I tried in OP's snack. Here is the modified snack (line 23):
https://snack.expo.io/B1-0GLCD4

**Comment by matthamil** [comment]
@toancaro if you are still experiencing issues, I recommend checking out [Reactiflux](https://www.reactiflux.com/) or Stack Overflow. Feel free to open a new issue if you believe this is a bug.
